### PR TITLE
Fix m4 to compile with Intel

### LINKED ISSUE #5705: openmpi fails with intel@17.0.1
```
$ spack install -j1 openmpi@2.1.1%intel@17.0.1
==> xz is already installed in /home/certik/repos/spack/opt/spack/linux-fedora25-x86_64/intel-17.0.1/xz-5.2.3-ajkczqbkmo76meocrthfrwnbup3aclzk
==> zlib is already installed in /home/certik/repos/spack/opt/spack/linux-fedora25-x86_64/intel-17.0.1/zlib-1.2.11-27anjzuze64nyci6djr45njj4t2wjdz3
==> pkg-config is already installed in /home/certik/repos/spack/opt/spack/linux-fedora25-x86_64/intel-17.0.1/pkg-config-0.29.2-iln5fu3hz6i5lrnmqerngwzsjz4a4zwu
==> Installing libxml2
==> Using cached archive: /home/certik/repos/spack/var/spack/cache/libxml2/libxml2-2.9.4.tar.gz
==> Staging archive: /home/certik/repos/spack/var/spack/stage/libxml2-2.9.4-7x7qw2zw4yzxue3dbpbp34tajpzsxl7m/libxml2-2.9.4.tar.gz
==> Created stage in /home/certik/repos/spack/var/spack/stage/libxml2-2.9.4-7x7qw2zw4yzxue3dbpbp34tajpzsxl7m
==> No patches needed for libxml2
==> Building libxml2 [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make'

4 errors found in build log:
     [ ... ]
     1147  icc: command line warning #10006: ignoring unknown option '-Wredundant-decls'
     1148  ld: warning: libimf.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
     1149  ld: warning: libsvml.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
     1150  ld: warning: libirng.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
     1151  ld: warning: libintlc.so.5, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
     1152  ld: .libs/xmllint: hidden symbol `__intel_cpu_features_init_x' in /opt/intel/compilers_and_libraries_2017.1.132/linux/compiler/lib/intel64_lin/libirc.a(cpu_feature_disp.o) is referenced by DSO
  >> 1153  ld: final link failed: Bad value
     1154  Makefile:1117: recipe for target 'xmllint' failed
  >> 1155  make[2]: *** [xmllint] Error 1
     1156  make[2]: Leaving directory '/tmp/certik/spack-stage/spack-stage-YzPtwr/libxml2-2.9.4'
     1157  Makefile:1440: recipe for target 'all-recursive' failed
  >> 1158  make[1]: *** [all-recursive] Error 1
     1159  make[1]: Leaving directory '/tmp/certik/spack-stage/spack-stage-YzPtwr/libxml2-2.9.4'
     1160  Makefile:859: recipe for target 'all' failed
  >> 1161  make: *** [all] Error 2

See build log for details:
```
The relevant part of the log is at the end:
```
  CCLD     testdso.la
  CC       xmllint.o
icc: command line warning #10006: ignoring unknown option '-Wno-format-extra-args'
icc: command line warning #10006: ignoring unknown option '-Wcast-align'
icc: command line warning #10006: ignoring unknown option '-Waggregate-return'
icc: command line warning #10006: ignoring unknown option '-Wnested-externs'
icc: command line warning #10006: ignoring unknown option '-Wredundant-decls'
  CCLD     xmllint
icc: command line warning #10006: ignoring unknown option '-Wno-format-extra-args'
icc: command line warning #10006: ignoring unknown option '-Wcast-align'
icc: command line warning #10006: ignoring unknown option '-Waggregate-return'
icc: command line warning #10006: ignoring unknown option '-Wnested-externs'
icc: command line warning #10006: ignoring unknown option '-Wredundant-decls'
ld: warning: libimf.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
ld: warning: libsvml.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
ld: warning: libirng.so, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
ld: warning: libintlc.so.5, needed by ./.libs/libxml2.so, not found (try using -rpath or -rpath-link)
ld: .libs/xmllint: hidden symbol `__intel_cpu_features_init_x' in /opt/intel/compilers_and_libraries_2017.1.132/linux/compiler/lib/intel64_lin/libirc.a(cpu_feature_disp.o) is referenced by DSO
ld: final link failed: Bad value
Makefile:1117: recipe for target 'xmllint' failed
make[2]: *** [xmllint] Error 1
make[2]: Leaving directory '/tmp/certik/spack-stage/spack-stage-YzPtwr/libxml2-2.9.4'
Makefile:1440: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/tmp/certik/spack-stage/spack-stage-YzPtwr/libxml2-2.9.4'
Makefile:859: recipe for target 'all' failed
make: *** [all] Error 2
```

### PR DESCRIPTION
**Pull request by certik**
Fixes #5705.

### CODE REVIEW

File: var/spack/repos/builtin/packages/m4/package.py [hunk 1]
```diff
@@ -51,6 +51,9 @@ def configure_args(self):
         if spec.satisfies('%clang') and not spec.satisfies('platform=darwin'):
             args.append('CFLAGS=-rtlib=compiler-rt')
 
+        if spec.satisfies('%intel'):
+            args.append('CFLAGS=-no-gcc')
```
Comment (junghans) [inline var/spack/repos/builtin/packages/m4/package.py]: Would that overwrite user defined `CFLAGS`?

Comment (certik) [reply var/spack/repos/builtin/packages/m4/package.py]: Yes, it feels like a hack. The same with the previous `clang` part that is already in the `develop` branch.

Comment (junghans) [reply var/spack/repos/builtin/packages/m4/package.py]: I tried `./configure CFLAGS="-O1 -g" CFLAGS="-O2"` on a different package and only `-O2` is picked up for building.

I think we want to use `cflags.append(-no-gcc)` and the same for the above flag.

Comment (certik) [reply var/spack/repos/builtin/packages/m4/package.py]: Fixed in 4cb1c2b1d7540a75d7b17312d2732087fbf94587.

Comment (certik) [reply var/spack/repos/builtin/packages/m4/package.py]: @junghans I don't see how the change you requested can work. I did try it, but spack doesn't pick up this change and says the package was already built. But I think the `cflags` is just a local variable that will not do anything.

Comment (certik) [reply var/spack/repos/builtin/packages/m4/package.py]: I will be off my computer. If you figure it out, feel free to push into this PR, you should have the rights.

**Comment by junghans** [comment]
@adamjstewart please review!

File: var/spack/repos/builtin/packages/m4/package.py [hunk 2]
```diff
@@ -49,7 +49,10 @@ def configure_args(self):
         args = ['--enable-c++']
 
         if spec.satisfies('%clang') and not spec.satisfies('platform=darwin'):
-            args.append('CFLAGS=-rtlib=compiler-rt')
+            cflags.append('-rtlib=compiler-rt')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/m4/package.py]: I know I've been away from Spack for a while, so perhaps things have changed internally, but... where is `cflags` defined?

Comment (certik) [reply var/spack/repos/builtin/packages/m4/package.py]: Yes, that's precisely what I asked in https://github.com/LLNL/spack/pull/5728#discussion_r144389378.

Comment (junghans) [reply var/spack/repos/builtin/packages/m4/package.py]: I screwed up, I looked at `cp2k` and forgot to add:
```
cflags = copy.deepcopy(optflags[self.spec.compiler.name])
```
plus
```
args.append('CFLAGS = {0}'.format(' '.join(cflags)))
```
Lame....

**Comment by junghans** [comment]
I dropped the last commit, let's use the original solution and fix it later if somebody needs user `CFLAGS` on m4.

**Comment by certik** [comment]
I literally *just* finished a commit with the correct cflags. Should I push it in?

**Comment by certik** [comment]
I rebased my branch and push it into #5756.

**Pull request merged**